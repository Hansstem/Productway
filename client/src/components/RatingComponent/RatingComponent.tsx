import { useState } from 'react';

import { Alert, Snackbar } from '@mui/material';
import Rating, { IconContainerProps } from '@mui/material/Rating';
import { styled } from '@mui/material/styles';

import customIcons from '../../data/ratingIcons.tsx';

/*
 * Styling for the rating icons
 */
const StyledRating = styled(Rating)(({ theme }) => ({
	'& .MuiRating-iconEmpty .MuiSvgIcon-root': {
		color: theme.palette.action.disabled,
	},
}));

/*
 * function for displaying the rating icons
 */
function IconContainer(props: IconContainerProps) {
	const { value, ...other } = props;
	return <span {...other}>{customIcons[value].icon}</span>;
}
/**
 * This interface contains a rating and a function for handling the rating change
 * @param {number} rating - The rating of the product
 * @param {function} onRatingChange - The function that handles the rating change
 */
interface RatingComponentProps {
	rating: number;
	onRatingChange: (
		event: React.SyntheticEvent<Element, Event>,
		newValue: number | null,
	) => void;
}
/**
 * This component is for displaying the rating icons
 * @param {RatingComponentProps} props - The rating and the function for handling the rating change
 */
export default function RatingComponent(props: RatingComponentProps) {
	const [rating, setRating] = useState(props.rating);
	const [rated, setRated] = useState(false);
	const [removeRating, setRemovedRating] = useState(false);

	const ratingChange = (
		_event: React.SyntheticEvent<Element, Event>,
		newValue: number | null,
	) => {
		setRating(newValue ? Number(newValue) : 0);
		if (newValue === 0 || newValue === null) {
			setRated(false);
			setRemovedRating(true);
			props.onRatingChange(_event, 0);
			return;
		} else {
			setRemovedRating(false);
			setRated(true);
			props.onRatingChange(_event, newValue);
		}
	};
	/*
	 * This function closes the snackbar after autoHideDuration.
	 */
	const handleClose = (
		_event?: React.SyntheticEvent | Event,
		reason?: string,
	) => {
		if (reason === 'clickaway') {
			return;
		}
		setRated(false);
		setRemovedRating(false);
	};

	return (
		<>
			<StyledRating
				name="highlight-selected-only"
				value={rating}
				IconContainerComponent={IconContainer}
				getLabelText={(value: number) => customIcons[value].label}
				highlightSelectedOnly
				onChange={(
					event: React.SyntheticEvent<Element, Event>,
					newValue: number | null,
				) => ratingChange(event, newValue)}
			/>
			<Snackbar
				open={rated}
				autoHideDuration={3000}
				onClose={handleClose}
			>
				<Alert
					onClose={handleClose}
					severity="success"
					sx={{ width: '100%' }}
				>
					Produkt vurdert til {rating}
				</Alert>
			</Snackbar>
			<Snackbar
				open={removeRating}
				autoHideDuration={3000}
				onClose={handleClose}
			>
				<Alert
					onClose={handleClose}
					severity="error"
					sx={{ width: '100%' }}
				>
					Fjernet vurdering!
				</Alert>
			</Snackbar>
		</>
	);
}
