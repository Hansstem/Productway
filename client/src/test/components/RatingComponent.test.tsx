import { render, RenderResult, screen } from '@testing-library/react';
import jest from 'jest-mock';

import RatingComponent from '../../components/RatingComponent/RatingComponent';

describe('RatingComponent test', () => {
	let ratingComponent: RenderResult;
	let mockRatingChange: jest.Mock;

	beforeEach(() => {
		ratingComponent = render(
			<>
				<RatingComponent rating={5} onRatingChange={mockRatingChange} />
			</>,
		);
	});

	it('snapshottest errormessage', () => {
		expect(ratingComponent).toMatchSnapshot();
	});

	it('renders correctly', () => {
		const ratingShit = screen.getByLabelText('Very Satisfied');
		expect(ratingShit).toBeInTheDocument();
	});
});
