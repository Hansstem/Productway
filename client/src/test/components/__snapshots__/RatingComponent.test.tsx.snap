// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`RatingComponent test > snapshottest errormessage 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <span
        class="MuiRating-root MuiRating-sizeMedium css-kglq79-MuiRating-root"
      >
        <label
          class="css-dqr9h-MuiRating-label"
          for=":r1:"
        >
          <span
            class="MuiRating-icon MuiRating-iconEmpty css-1c99szj-MuiRating-icon"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-colorError MuiSvgIcon-fontSizeLarge css-1o657us-MuiSvgIcon-root"
              data-testid="SentimentVeryDissatisfiedIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <circle
                cx="15.5"
                cy="9.5"
                r="1.5"
              />
              <circle
                cx="8.5"
                cy="9.5"
                r="1.5"
              />
              <path
                d="M11.99 2C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8zm0-6c-2.33 0-4.32 1.45-5.12 3.5h1.67c.69-1.19 1.97-2 3.45-2s2.75.81 3.45 2h1.67c-.8-2.05-2.79-3.5-5.12-3.5z"
              />
            </svg>
          </span>
          <span
            class="MuiRating-visuallyHidden"
          >
            Very Dissatisfied
          </span>
        </label>
        <input
          class="MuiRating-visuallyHidden"
          id=":r1:"
          name="highlight-selected-only"
          type="radio"
          value="1"
        />
        <label
          class="css-dqr9h-MuiRating-label"
          for=":r2:"
        >
          <span
            class="MuiRating-icon MuiRating-iconEmpty css-1c99szj-MuiRating-icon"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-colorError MuiSvgIcon-fontSizeLarge css-1o657us-MuiSvgIcon-root"
              data-testid="SentimentDissatisfiedIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <circle
                cx="15.5"
                cy="9.5"
                r="1.5"
              />
              <circle
                cx="8.5"
                cy="9.5"
                r="1.5"
              />
              <path
                d="M11.99 2C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8zm0-3.5c.73 0 1.39.19 1.97.53.12-.14.86-.98 1.01-1.14-.85-.56-1.87-.89-2.98-.89-1.11 0-2.13.33-2.99.88.97 1.09.01.02 1.01 1.14.59-.33 1.25-.52 1.98-.52z"
              />
            </svg>
          </span>
          <span
            class="MuiRating-visuallyHidden"
          >
            Dissatisfied
          </span>
        </label>
        <input
          class="MuiRating-visuallyHidden"
          id=":r2:"
          name="highlight-selected-only"
          type="radio"
          value="2"
        />
        <label
          class="css-dqr9h-MuiRating-label"
          for=":r3:"
        >
          <span
            class="MuiRating-icon MuiRating-iconEmpty css-1c99szj-MuiRating-icon"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-colorWarning MuiSvgIcon-fontSizeLarge css-1uo3q5b-MuiSvgIcon-root"
              data-testid="SentimentSatisfiedIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <circle
                cx="15.5"
                cy="9.5"
                r="1.5"
              />
              <circle
                cx="8.5"
                cy="9.5"
                r="1.5"
              />
              <path
                d="M11.99 2C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8zm0-4c-.73 0-1.38-.18-1.96-.52-.12.14-.86.98-1.01 1.15.86.55 1.87.87 2.97.87 1.11 0 2.12-.33 2.98-.88-.97-1.09-.01-.02-1.01-1.15-.59.35-1.24.53-1.97.53z"
              />
            </svg>
          </span>
          <span
            class="MuiRating-visuallyHidden"
          >
            Neutral
          </span>
        </label>
        <input
          class="MuiRating-visuallyHidden"
          id=":r3:"
          name="highlight-selected-only"
          type="radio"
          value="3"
        />
        <label
          class="css-dqr9h-MuiRating-label"
          for=":r4:"
        >
          <span
            class="MuiRating-icon MuiRating-iconEmpty css-1c99szj-MuiRating-icon"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-colorSuccess MuiSvgIcon-fontSizeLarge css-1mhxotf-MuiSvgIcon-root"
              data-testid="SentimentSatisfiedAltOutlinedIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <circle
                cx="15.5"
                cy="9.5"
                r="1.5"
              />
              <circle
                cx="8.5"
                cy="9.5"
                r="1.5"
              />
              <path
                d="M12 16c-1.48 0-2.75-.81-3.45-2H6.88c.8 2.05 2.79 3.5 5.12 3.5s4.32-1.45 5.12-3.5h-1.67c-.69 1.19-1.97 2-3.45 2zm-.01-14C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
              />
            </svg>
          </span>
          <span
            class="MuiRating-visuallyHidden"
          >
            Satisfied
          </span>
        </label>
        <input
          class="MuiRating-visuallyHidden"
          id=":r4:"
          name="highlight-selected-only"
          type="radio"
          value="4"
        />
        <label
          class="css-dqr9h-MuiRating-label"
          for=":r5:"
        >
          <span
            class="MuiRating-icon MuiRating-iconFilled css-7qmtgc-MuiRating-icon"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-colorSuccess MuiSvgIcon-fontSizeLarge css-1mhxotf-MuiSvgIcon-root"
              data-testid="SentimentVerySatisfiedIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <circle
                cx="15.5"
                cy="9.5"
                r="1.5"
              />
              <circle
                cx="8.5"
                cy="9.5"
                r="1.5"
              />
              <path
                d="M11.99 2C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8zm-5-6c.78 2.34 2.72 4 5 4s4.22-1.66 5-4H7z"
              />
            </svg>
          </span>
          <span
            class="MuiRating-visuallyHidden"
          >
            Very Satisfied
          </span>
        </label>
        <input
          checked=""
          class="MuiRating-visuallyHidden"
          id=":r5:"
          name="highlight-selected-only"
          type="radio"
          value="5"
        />
        <label
          class="MuiRating-label MuiRating-pristine css-dqr9h-MuiRating-label"
        >
          <input
            class="MuiRating-visuallyHidden"
            id="highlight-selected-only-empty"
            name="highlight-selected-only"
            type="radio"
            value=""
          />
          <span
            class="MuiRating-visuallyHidden"
          >
            Empty
          </span>
        </label>
      </span>
    </div>
  </body>,
  "container": <div>
    <span
      class="MuiRating-root MuiRating-sizeMedium css-kglq79-MuiRating-root"
    >
      <label
        class="css-dqr9h-MuiRating-label"
        for=":r1:"
      >
        <span
          class="MuiRating-icon MuiRating-iconEmpty css-1c99szj-MuiRating-icon"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-colorError MuiSvgIcon-fontSizeLarge css-1o657us-MuiSvgIcon-root"
            data-testid="SentimentVeryDissatisfiedIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <circle
              cx="15.5"
              cy="9.5"
              r="1.5"
            />
            <circle
              cx="8.5"
              cy="9.5"
              r="1.5"
            />
            <path
              d="M11.99 2C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8zm0-6c-2.33 0-4.32 1.45-5.12 3.5h1.67c.69-1.19 1.97-2 3.45-2s2.75.81 3.45 2h1.67c-.8-2.05-2.79-3.5-5.12-3.5z"
            />
          </svg>
        </span>
        <span
          class="MuiRating-visuallyHidden"
        >
          Very Dissatisfied
        </span>
      </label>
      <input
        class="MuiRating-visuallyHidden"
        id=":r1:"
        name="highlight-selected-only"
        type="radio"
        value="1"
      />
      <label
        class="css-dqr9h-MuiRating-label"
        for=":r2:"
      >
        <span
          class="MuiRating-icon MuiRating-iconEmpty css-1c99szj-MuiRating-icon"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-colorError MuiSvgIcon-fontSizeLarge css-1o657us-MuiSvgIcon-root"
            data-testid="SentimentDissatisfiedIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <circle
              cx="15.5"
              cy="9.5"
              r="1.5"
            />
            <circle
              cx="8.5"
              cy="9.5"
              r="1.5"
            />
            <path
              d="M11.99 2C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8zm0-3.5c.73 0 1.39.19 1.97.53.12-.14.86-.98 1.01-1.14-.85-.56-1.87-.89-2.98-.89-1.11 0-2.13.33-2.99.88.97 1.09.01.02 1.01 1.14.59-.33 1.25-.52 1.98-.52z"
            />
          </svg>
        </span>
        <span
          class="MuiRating-visuallyHidden"
        >
          Dissatisfied
        </span>
      </label>
      <input
        class="MuiRating-visuallyHidden"
        id=":r2:"
        name="highlight-selected-only"
        type="radio"
        value="2"
      />
      <label
        class="css-dqr9h-MuiRating-label"
        for=":r3:"
      >
        <span
          class="MuiRating-icon MuiRating-iconEmpty css-1c99szj-MuiRating-icon"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-colorWarning MuiSvgIcon-fontSizeLarge css-1uo3q5b-MuiSvgIcon-root"
            data-testid="SentimentSatisfiedIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <circle
              cx="15.5"
              cy="9.5"
              r="1.5"
            />
            <circle
              cx="8.5"
              cy="9.5"
              r="1.5"
            />
            <path
              d="M11.99 2C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8zm0-4c-.73 0-1.38-.18-1.96-.52-.12.14-.86.98-1.01 1.15.86.55 1.87.87 2.97.87 1.11 0 2.12-.33 2.98-.88-.97-1.09-.01-.02-1.01-1.15-.59.35-1.24.53-1.97.53z"
            />
          </svg>
        </span>
        <span
          class="MuiRating-visuallyHidden"
        >
          Neutral
        </span>
      </label>
      <input
        class="MuiRating-visuallyHidden"
        id=":r3:"
        name="highlight-selected-only"
        type="radio"
        value="3"
      />
      <label
        class="css-dqr9h-MuiRating-label"
        for=":r4:"
      >
        <span
          class="MuiRating-icon MuiRating-iconEmpty css-1c99szj-MuiRating-icon"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-colorSuccess MuiSvgIcon-fontSizeLarge css-1mhxotf-MuiSvgIcon-root"
            data-testid="SentimentSatisfiedAltOutlinedIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <circle
              cx="15.5"
              cy="9.5"
              r="1.5"
            />
            <circle
              cx="8.5"
              cy="9.5"
              r="1.5"
            />
            <path
              d="M12 16c-1.48 0-2.75-.81-3.45-2H6.88c.8 2.05 2.79 3.5 5.12 3.5s4.32-1.45 5.12-3.5h-1.67c-.69 1.19-1.97 2-3.45 2zm-.01-14C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
            />
          </svg>
        </span>
        <span
          class="MuiRating-visuallyHidden"
        >
          Satisfied
        </span>
      </label>
      <input
        class="MuiRating-visuallyHidden"
        id=":r4:"
        name="highlight-selected-only"
        type="radio"
        value="4"
      />
      <label
        class="css-dqr9h-MuiRating-label"
        for=":r5:"
      >
        <span
          class="MuiRating-icon MuiRating-iconFilled css-7qmtgc-MuiRating-icon"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-colorSuccess MuiSvgIcon-fontSizeLarge css-1mhxotf-MuiSvgIcon-root"
            data-testid="SentimentVerySatisfiedIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <circle
              cx="15.5"
              cy="9.5"
              r="1.5"
            />
            <circle
              cx="8.5"
              cy="9.5"
              r="1.5"
            />
            <path
              d="M11.99 2C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8zm-5-6c.78 2.34 2.72 4 5 4s4.22-1.66 5-4H7z"
            />
          </svg>
        </span>
        <span
          class="MuiRating-visuallyHidden"
        >
          Very Satisfied
        </span>
      </label>
      <input
        checked=""
        class="MuiRating-visuallyHidden"
        id=":r5:"
        name="highlight-selected-only"
        type="radio"
        value="5"
      />
      <label
        class="MuiRating-label MuiRating-pristine css-dqr9h-MuiRating-label"
      >
        <input
          class="MuiRating-visuallyHidden"
          id="highlight-selected-only-empty"
          name="highlight-selected-only"
          type="radio"
          value=""
        />
        <span
          class="MuiRating-visuallyHidden"
        >
          Empty
        </span>
      </label>
    </span>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
